### DISCUSSION TITLE
How to update the cards in the drawer?

### DISCUSSION THREAD

**Discussion by Yao-14** [Kitware/trame#225] [Q&A] [answered]
Sorry to bother you again, I tried a method similar to gittree update to update the card, but it has not been a good result. Can you tell me a way to update cards similar to gittree update? My need is similar to gittree update, when I convert the model, all the cards can be updated at the same time.

**Comment by Yao-14** [comment]
```
def pipeline(server, plotter):
    """Create a vuetify GitTree."""
    state, ctrl = server.state, server.controller

    # Selection Change
    @ctrl.set("actives_change")
    def actives_change(ids):
        _id = ids[0]
        active_actor_id = state.actor_ids[int(_id) - 1]
        state.active_ui = active_actor_id
        state.active_id = int(_id)
        ctrl.view_update()

    # Visibility Change
    @ctrl.set("visibility_change")
    def visibility_change(event):
        _id = event["id"]
        _visibility = event["visible"]
        active_actor = [value for value in plotter.actors.values()][int(_id) - 1]
        active_actor.SetVisibility(_visibility)
        ctrl.view_update()

    trame.GitTree(
        sources=("pipeline",),
        actives_change=(ctrl.actives_change, "[$event]"),
        visibility_change=(ctrl.visibility_change, "[$event]"),
    )

    with vuetify.VCard(): # v_show="active_ui"
        vuetify.VCardTitle(
            "{{ active_ui }}",
            classes="grey lighten-1 py-1 grey--text text--darken-3",
            style="user-select: none; cursor: pointer",
            hide_details=True,
            dense=True,
        )
        actors = [value for value in plotter.actors.values()]
        for actor, actor_id in zip(actors, server.state.actor_ids):
            with vuetify.VCardText(classes="py-2", v_show=f"active_ui === '{actor_id}'"):
                CBinCard = PVCB(server=server, actor=actor, actor_id=actor_id)
                if str(actor_id).startswith("PC"):
                    standard_pc_card(CBinCard)
                if str(actor_id).startswith("Mesh"):
                    standard_mesh_card(CBinCard)
```

**Comment by jourdain** [comment]
`for actor, actor_id in zip(actors, server.state.actor_ids):` will be evaluated only once when the UI get created. If the list of actors are not ready yet, there will be no UI generated at all.

**Comment by Yao-14** [comment]
Yes, I know this problem exists, is there any good alternative?

**Comment by jourdain** [comment]
Yes you just build the UI once using shared/global variable names that you use to perform action on the "active" element. 

So when the active change, you update the variables that display things in your vcards.
When user change things, you propagate them to the active item.

As simple as that.

**Comment by Yao-14** [comment]
Thank you very much, your solution is very attractive, but I don't know exactly what to change. Could you please give me a concrete example?

**Comment by jourdain** [comment, marked as answer]
Don't do a [for loop](https://github.com/aristoteleo/spateo-viewer/blob/main/stviewer/ui/drawer.py#L194). Just run it once.
Then use a [variable name](https://github.com/aristoteleo/spateo-viewer/blob/main/stviewer/ui/drawer.py#L159) that is fixed.

Then on active actor change, update all those fixed vars using the value available in the active actor.
And when those fixed vars change, you update the active actor with what they are modifying.

**Comment by Yao-14** [comment]
Great, after trial and error, I managed to implement these functions. Thanks a lot for your help.

**Comment by Yao-14** [comment]
This tool will be published concurrently as part of our paper, and any help during this time is greatly appreciated.
